blocks/graph.py

```diff
@@ -32,45 +59,58 @@ def __init__(self, outputs):
-                application_call = current.tag.application_call
-                if application_call not in self.application_calls:
-                    self.application_calls.add(application_call)
-                    for av in application_call.auxiliary_variables:
-                        av.tag.application_call = current.tag.application_call
-                        recursion(av)
-                    self.updates.extend(application_call.updates)
-            if current.owner:
-                owner = current.owner
-                if owner not in self.applies:
-                    if hasattr(owner.tag, 'updates'):
-                        logger.debug("found updates in application of {}"
-                                     .format(owner))
-                        self.updates.extend(owner.tag.updates.items())
-                    self.applies.add(owner)
-                for input_ in owner.inputs:
-                    if input_ not in self.variables:
-                        recursion(input_)
-
-        for output in self.outputs:
-            if output not in self.variables:
-                recursion(output)
-        self.inputs = [v for v in self.variables if is_graph_input(v)]
-
-    def get_shared_variables(self):
-        """Returns all shared variables found in the computation graph."""
+    @property
+    def inputs(self):
+        """Inputs to the graph, excluding constants and shared variables."""
+        return [var for var in self.variables if is_graph_input(var)]
+
+    @property
+    def intermediary_variables(self):
+        return [var for var in self.variables if
+                var not in self.inputs and
+                var not in self.outputs]
+
+    @property
+    def shared_variables(self):
         return [var for var in self.variables if is_shared_variable(var)]
 
+    @property
+    def auxiliary_variables(self):
+        return [var for var in self.variables if hasattr(var.tag, 'roles') and
+                VariableRole.AUXILIARY in var.tag.roles]
+
+    def _get_variables(self):
+        """Collect variables, updates and auxiliary variables."""
+        updates = OrderedDict()
+
+        # Sort apply nodes topologically, get variables and remove duplicates
+        inputs = graph.inputs(self.outputs)
+        sorted_apply_nodes = sort_apply_nodes([inputs], self.outputs,
+                                              [dependence])
+        seen = set()
+        variables = [var for var in list(chain(
+            *[apply_node.inputs for apply_node in sorted_apply_nodes]))
+            if not (var in seen or seen.add(var))] + self.outputs
+
+        # While preserving order add auxiliary variables, and collect updates
+        i = 0
+        seen = {'application_call': set(), 'annotation': set(), 'brick': set()}
+        while i < len(variables):
+            var = variables[i]
+            for tag in ('application_call', 'annotation', 'brick'):
```

Comment:
This is not exactly what we agreed on in #146. In the comment https://github.com/bartvm/blocks/issues/146#issuecomment-71247985 I meant the following: every variable has a single annotation: `x.tag.annotation`. 
This annotation can be attached manually, bust most often it is attached automatically in `Application.apply`. In this case 
the annotation is an instance of `ApplicationCall`, which subclasses `Application`. The annotation attached in `Application.apply` consists of variables put there by the user and the auxiliary variables of the brick this application belongs to.

I just want to keep the computation graph code clean from brick-related concepts. If it is a thing on its own it will be easier to maintain and keep documented.

Perhaps a better way to implement what is described above: let a variable have several annotations just they currently can have several roles and just as in your PR, with the distinction that all of them belongs to `x.tag.annotations` list. Then in `Application.apply` simply two annotations can be added: one from the brick and another one just an `ApplicationCall` instance.